python-stdlib/logging/logging.py

```diff
@@ -1,128 +1,202 @@
 
-    def _level_str(self, level):
-        l = _level_dict.get(level)
-        if l is not None:
-            return l
-        return "LVL%s" % level
+
+class Logger:
+    def __init__(self, name):
+        self.name = name
+        self.level = NOTSET
+        self.handlers = []
+        self.record = Record()
 
     def setLevel(self, level):
         self.level = level
 
-    def isEnabledFor(self, level):
-        return level >= (self.level or _level)
+    def addHandler(self, handler):
+        self.handlers.append(handler)
 
-    def log(self, level, msg, *args):
-        if self.isEnabledFor(level):
-            levelname = self._level_str(level)
-            if args:
-                msg = msg % args
-            if self.handlers:
-                d = self.record.__dict__
-                d["levelname"] = levelname
-                d["levelno"] = level
-                d["message"] = msg
-                d["name"] = self.name
-                for h in self.handlers:
-                    h.emit(self.record)
-            else:
-                print(levelname, ":", self.name, ":", msg, sep="", file=_stream)
+    def hasHandlers(self):
+        return len(self.handlers) > 0
 
-    def debug(self, msg, *args):
-        self.log(DEBUG, msg, *args)
+    def debug(self, message, *args, **kwargs):
+        self.log(DEBUG, message, *args, **kwargs)
 
-    def info(self, msg, *args):
-        self.log(INFO, msg, *args)
+    def info(self, message, *args, **kwargs):
+        self.log(INFO, message, *args, **kwargs)
 
-    def warning(self, msg, *args):
-        self.log(WARNING, msg, *args)
+    def warning(self, message, *args, **kwargs):
+        self.log(WARNING, message, *args, **kwargs)
 
-    def error(self, msg, *args):
-        self.log(ERROR, msg, *args)
+    def error(self, message, *args, **kwargs):
+        self.log(ERROR, message, *args, **kwargs)
 
-    def critical(self, msg, *args):
-        self.log(CRITICAL, msg, *args)
+    def critical(self, message, *args, **kwargs):
+        self.log(CRITICAL, message, *args, **kwargs)
 
-    def exc(self, e, msg, *args):
-        self.log(ERROR, msg, *args)
-        sys.print_exception(e, _stream)
+    def exception(self, message, *args, **kwargs):
+        self.log(ERROR, message, *args, **kwargs)
+        if hasattr(sys, "exc_info"):
+            for h in filter(lambda h: isinstance(h, StreamHandler), self.handlers):
+                sys.print_exception(sys.exc_info()[1], h.stream)
 
-    def exception(self, msg, *args):
-        self.exc(sys.exc_info()[1], msg, *args)
+    def log(self, level, message, *args, **kwargs):
+        if level >= self.level:
+            if args and isinstance(args[0], dict):
+                args = args[0]
+            for h in self.handlers:
+                self.record.set(self.name, level, message % args)
+                h.emit(self.record)
 
-    def addHandler(self, hndlr):
-        self.handlers.append(hndlr)
 
+def debug(message, *args, **kwargs):
```

Comment:
Do we need to support `**kwargs`?  It looks like they are currently unused, and having them here means each call to this function creates a new dict (which will be empty if no kwargs are passed in).

Reply:
Removed.